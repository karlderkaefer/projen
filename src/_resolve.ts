import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
