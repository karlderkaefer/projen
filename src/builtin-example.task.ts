console.log("hello, I am a builtin task");
console.log("second line");
