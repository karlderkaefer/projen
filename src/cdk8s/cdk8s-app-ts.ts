import * as path from "path";
import * as fs from "fs-extra";
import { Component } from "../component";
import { TypeScriptAppProject, TypeScriptProjectOptions } from "../typescript";
import { YamlFile } from "../yaml";
import { AutoDiscover } from "./auto-discover";

export interface Cdk8sTypeScriptAppOptions extends TypeScriptProjectOptions {
  /**
   * Minimum target version this library is tested against.
   *
   * @default "1.0.0-beta.10"
   * @featured
   */
  readonly cdk8sVersion: string;

  /**
   * Import a specific Kubernetes spec version.
   *
   * @default - Use the cdk8s default
   */
  readonly k8sSpecVersion?: string;

  /**
   * Import additional specs
   *
   * @default - no additional specs imported
   */
  readonly cdk8sImports?: string[];

  /**
   * constructs verion
   *
   * @default "3.2.34"
   */

  readonly constructsVersion?: string;

  /**
   * cdk8s-cli version
   *
   * @default "cdk8sVersion"
   */

  readonly cdk8sCliVersion?: string;

  /**
   * Use pinned version instead of caret version for CDK8s.
   *
   * You can use this to prevent yarn to mix versions for your CDK8s package and to prevent auto-updates.
   * If you use experimental features this will let you define the moment you include breaking changes.
   *
   * @default false
   */
  readonly cdk8sVersionPinning?: boolean;

  /**
   * Use pinned version instead of caret version for CDK8s-cli.
   *
   * You can use this to prevent yarn to mix versions for your CDK8s package and to prevent auto-updates.
   * If you use experimental features this will let you define the moment you include breaking changes.
   *
   * @default false
   */
  readonly cdk8sCliVersionPinning?: boolean;

  /**
   * Use pinned version instead of caret version for cdk8s-plus-17.
   *
   * You can use this to prevent yarn to mix versions for your CDK8s package and to prevent auto-updates.
   * If you use experimental features this will let you define the moment you include breaking changes.
   *
   * @default false
   */
  readonly cdk8sPlusVersionPinning?: boolean;

  /**
   * Use pinned version instead of caret version for constructs.
   *
   * You can use this to prevent yarn to mix versions for your consructs package and to prevent auto-updates.
   * If you use experimental features this will let you define the moment you include breaking changes.
   *
   * @default false
   */
  readonly constructsVersionPinning?: boolean;

  /**
   * The CDK8s app's entrypoint (relative to the source directory, which is
   * "src" by default).
   *
   * @default "main.ts"
   */
  readonly appEntrypoint?: string;

  /**
   * Automatically adds an `cdk8s.IntegrationTest` for each `.integ.ts` app
   * in your test directory. If this is disabled, you can manually add an
   * `cdk8s.AutoDiscover` component to your project.
   *
   * @default true
   */
  readonly integrationTestAutoDiscover?: boolean;
}

/**
 * CDK8s app in TypeScript
 *
 *
 * @pjid cdk8s-app-ts
 */

export class Cdk8sTypeScriptApp extends TypeScriptAppProject {
  /**
   * The CDK8s version this app is using.
   */
  public readonly cdk8sVersion: string;

  /**
   * The constructs version this app is using.
   */
  public readonly constructsVersion: string;

  /**
   * The cdk8s-cli version this app is using.
   */

  public readonly cdk8sCliVersion: string;

  /**
   * The CDK8s app entrypoint
   */
  public readonly appEntrypoint: string;

  constructor(options: Cdk8sTypeScriptAppOptions) {
    super({
      ...options,
      sampleCode: false,
    });

    if (!options.cdk8sVersion) {
      throw new Error("Required field cdk8sVersion is not specified.");
    }

    // encode a hidden assumption further down the chain
    if (this.srcdir !== "src") {
      throw new Error('sources are expected under the "src" directory');
    }

    // encode a hidden assumption further down the chain
    if (this.testdir !== "test") {
      throw new Error('test sources are expected under the "test" directory');
    }

    this.appEntrypoint = options.appEntrypoint ?? "main.ts";

    this.cdk8sVersion = options.cdk8sVersionPinning
      ? options.cdk8sVersion
      : `^${options.cdk8sVersion}`;

    if (options.constructsVersion) {
      this.constructsVersion = options.constructsVersionPinning
        ? options.constructsVersion
        : `^${options.constructsVersion}`;
    } else {
      this.constructsVersion = "^3.2.34";
    }

    if (!!options.cdk8sCliVersion) {
      this.cdk8sCliVersion = options.cdk8sCliVersionPinning
        ? options.cdk8sCliVersion
        : `^${options.cdk8sCliVersion}`;
    } else {
      this.cdk8sCliVersion = this.cdk8sVersion;
    }

    // CLI
    this.addDeps(
      `cdk8s@${this.cdk8sVersion}`,
      `constructs@${this.constructsVersion}`
    );
    this.addDevDeps(
      "ts-node",
      `cdk8s-cli@${this.cdk8sCliVersion}`,
      `cdk8s@${this.cdk8sVersion}`,
      `constructs@${this.constructsVersion}`
    );

    const synth = this.addTask("synth", {
      description:
        'Synthesizes your cdk8s app into dist (part of "yarn build")',
      exec: "cdk8s synth",
    });

    this.addTask("import", {
      description: "Imports API objects to your app by generating constructs.",
      exec: "cdk8s import -o src/imports",
    });

    // add synth to the build
    this.postCompileTask.spawn(synth);

    const cdk8sImports = options.cdk8sImports ?? [];
    const k8sSpec = options.k8sSpecVersion
      ? `k8s@${options.k8sSpecVersion}`
      : "k8s";

    const appEntrypointBaseName = path.basename(this.appEntrypoint, ".ts");

    new YamlFile(this, "cdk8s.yaml", {
      committed: true,
      editGitignore: true,
      obj: {
        language: "typescript",
        app: `node lib/${appEntrypointBaseName}.js`,
        imports: [k8sSpec, ...cdk8sImports],
      },
    });

    if (options.sampleCode ?? true) {
      new SampleCode(this);
    }

    new AutoDiscover(this, {
      testdir: this.testdir,
      tsconfigPath: this.tsconfigDev.fileName,
      integrationTestAutoDiscover: options.integrationTestAutoDiscover ?? true,
    });
  }
}

class SampleCode extends Component {
  private readonly appProject: Cdk8sTypeScriptApp;
  constructor(project: Cdk8sTypeScriptApp) {
    super(project);
    this.appProject = project;
  }

  public synthesize() {
    const outdir = this.project.outdir;
    const srcdir = path.join(outdir, this.appProject.srcdir);
    if (
      fs.pathExistsSync(srcdir) &&
      fs.readdirSync(srcdir).filter((x) => x.endsWith(".ts"))
    ) {
      return;
    }

    const srcCode = `import { Construct } from 'constructs';
import { App, Chart, ChartProps, ApiObject } from 'cdk8s';

export class MyChart extends Chart {
  constructor(scope: Construct, id: string, props: ChartProps = { }) {
    super(scope, id, props);

    const label = { app: 'hello-k8s' };



    new ApiObject(this, 'deployment', {
      apiVersion: "v1",
      kind: "Pod",
      metadata: {
          namespace: "frontend",
          name: "nginx",
          labels: label,
      },
      spec: {
          containers: [{
              name: "nginx",
              image: "nginx:1.14-alpine",
              resources: {
                  limits: {
                      memory: "20Mi",
                      cpu: 0.2,
                  },
              },
          }],
      },
    });
  }
}

const app = new App();
new MyChart(app, 'hello');
app.synth();`;

    fs.mkdirpSync(srcdir);
    fs.writeFileSync(path.join(srcdir, this.appProject.appEntrypoint), srcCode);
  }
}
