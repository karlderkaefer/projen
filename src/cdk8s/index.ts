export * from "./auto-discover";
export * from "./cdk8s-app-ts";
export * from "./cdk8s-construct";
export * from "./integration-test";
