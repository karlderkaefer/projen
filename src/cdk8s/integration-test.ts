import { join } from "path";
import { IntegrationTestBaseOptions, IntegrationTestBase } from "../cdk";
import { DependencyType } from "../dependencies";
import { Project } from "../project";

/**
 * Options for IntegrationTest
 */
export interface IntegrationTestOptions extends IntegrationTestBaseOptions {}

/**
 * CDK8S integration test.
 */
export class IntegrationTest extends IntegrationTestBase {
  constructor(project: Project, options: IntegrationTestOptions) {
    super(project, options);

    if (!project.deps.tryGetDependency("cdk8s-cli")) {
      project.deps.addDependency(`cdk8s-cli`, DependencyType.BUILD);
    }

    if (!project.deps.tryGetDependency("ts-node")) {
      project.deps.addDependency("ts-node", DependencyType.BUILD);
    }

    const deployDir = join(this.tmpDir, "deploy.cdk.out");
    const assertDir = join(this.tmpDir, "assert.cdk.out");
    const app = `ts-node -P ${options.tsconfigPath} ${options.entrypoint}`;

    // Synth, deploy, and capture the snapshot
    this.deployTask.exec(`rm -fr ${deployDir}`);
    this.deployTask.exec(`cdk8s synth --app "${app}" -o ${deployDir}`);
    this.deployTask.exec(`kubectl apply -f ${deployDir}`);
    // If deployment was successful, copy the deploy dir to the expected dir
    this.deployTask.exec(`rm -fr ${this.snapshotDir}`);
    this.deployTask.exec(`mv ${deployDir} ${this.snapshotDir}`);

    // Run a snapshot
    this.snapshotTask.exec(`rm -fr ${this.snapshotDir}`);
    this.snapshotTask.exec(`cdk8s synth --app "${app}" -o ${this.snapshotDir}`);

    // Assert that the snapshot has not changed (run during tests)
    this.assertTask.exec(
      `cdk8s synth --app "${app}" -o ${assertDir} > /dev/null`
    );
    this.assertTask.exec(`diff ${this.snapshotDir}/ ${assertDir}/`);
  }
}
