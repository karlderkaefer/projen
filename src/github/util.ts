export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}
