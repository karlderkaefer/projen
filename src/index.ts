export * from './awscdk-app-ts';
export * from './awscdk-construct';
export * from './cdk8s-app-ts';
export * from './cdk8s-construct';
export * from './cdktf-construct';
export * from './component';
export * from './construct-lib';
export * from './dev-env';
export * from './docker-compose';
export * from './eslint';
export * from './file';
export * from './gitpod';
export * from './ignore-file';
export * from './ini';
export * from './jest';
export * from './jsii-project';
export * from './json';
export * from './logger';
export * from './license';
export * from './makefile';
export * from './node-package';
export * from './node-project';
export * from './object-file';
export * from './option-hints';
export * from './project';
export * from './readme';
export * from './sample-file';
export * from './semver';
export * from './textfile';
export * from './toml';
export * from './typescript';
export * from './typescript-config';
export * from './version';
export * from './yaml';
export * from './xmlfile';
export * from './source-code';
export * from './upgrade-dependencies';
export * from './circle-ci';

// export submobules
export * as tasks from './tasks';
export * as web from './web';
export * as github from './github';
export * as vscode from './vscode';
export * as deps from './deps';
export * as json from './json/index'; // <-- we also have json.ts
export * as java from './java';
export * as python from './python';
export * as typescript from './typescript/index'; // <-- we also have typescript.ts
export * as javascript from './javascript';
export * as release from './release';
export * as git from './git';
