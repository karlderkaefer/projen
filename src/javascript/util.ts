import { basename, dirname, extname, join, sep } from "path";

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return join(dir, base);
}
