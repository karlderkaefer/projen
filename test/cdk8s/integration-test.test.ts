import { Cdk8sTypeScriptApp, IntegrationTest } from "../../src/cdk8s";
import { synthSnapshot } from "../util";

describe("IntegrationTest", () => {
  test("adding an integration test", () => {
    const project = new Cdk8sTypeScriptApp({
      cdk8sVersion: "1.0.0-beta.11",
      name: "project",
      defaultReleaseBranch: "main",
      releaseWorkflow: true,
    });

    // WHEN
    new IntegrationTest(project, {
      entrypoint: "test/my-test-name.integ.ts",
      tsconfigPath: project.tsconfigDev.fileName,
    });

    // THEN
    const output = synthSnapshot(project);
    const tasks = output[".projen/tasks.json"].tasks;
    expect(tasks["integ:my-test-name:snapshot"]).toMatchSnapshot();
    expect(tasks["integ:my-test-name:deploy"]).toMatchSnapshot();
    expect(tasks["integ:my-test-name:assert"]).toMatchSnapshot();
    expect(tasks["integ:snapshot-all"]).toMatchSnapshot();
    expect(tasks.test).toMatchSnapshot();
  });

  test("explicit name", () => {
    const project = new Cdk8sTypeScriptApp({
      cdk8sVersion: "1.0.0-beta.11",
      name: "project",
      defaultReleaseBranch: "main",
      releaseWorkflow: true,
    });

    // WHEN
    new IntegrationTest(project, {
      name: "foobar",
      entrypoint: "test/my-test-name.integ.ts",
      tsconfigPath: project.tsconfigDev.fileName,
    });

    // THEN
    const output = synthSnapshot(project);
    const tasks = output[".projen/tasks.json"].tasks;
    expect(tasks["integ:foobar:snapshot"]).toBeDefined();
    expect(tasks["integ:foobar:deploy"]).toBeDefined();
    expect(tasks["integ:foobar:assert"]).toBeDefined();
    expect(tasks["integ:snapshot-all"]).toMatchSnapshot();
    expect(tasks.test).toMatchSnapshot();
  });
});
