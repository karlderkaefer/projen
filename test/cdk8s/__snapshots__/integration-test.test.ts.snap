// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntegrationTest adding an integration test 1`] = `
Object {
  "description": "update snapshot for integration test \\"my-test-name\\"",
  "name": "integ:my-test-name:snapshot",
  "steps": Array [
    Object {
      "exec": "rm -fr test/my-test-name.integ.snapshot",
    },
    Object {
      "exec": "cdk8s synth --app \\"ts-node -P tsconfig.dev.json test/my-test-name.integ.ts\\" -o test/my-test-name.integ.snapshot",
    },
  ],
}
`;

exports[`IntegrationTest adding an integration test 2`] = `
Object {
  "description": "deploy integration test 'my-test-name' and capture snapshot",
  "name": "integ:my-test-name:deploy",
  "steps": Array [
    Object {
      "exec": "rm -fr test/.tmp/my-test-name.integ/deploy.cdk.out",
    },
    Object {
      "exec": "cdk8s synth --app \\"ts-node -P tsconfig.dev.json test/my-test-name.integ.ts\\" -o test/.tmp/my-test-name.integ/deploy.cdk.out",
    },
    Object {
      "exec": "kubectl apply -f test/.tmp/my-test-name.integ/deploy.cdk.out",
    },
    Object {
      "exec": "rm -fr test/my-test-name.integ.snapshot",
    },
    Object {
      "exec": "mv test/.tmp/my-test-name.integ/deploy.cdk.out test/my-test-name.integ.snapshot",
    },
  ],
}
`;

exports[`IntegrationTest adding an integration test 3`] = `
Object {
  "description": "assert the snapshot of integration test 'my-test-name'",
  "name": "integ:my-test-name:assert",
  "steps": Array [
    Object {
      "exec": "[ -d \\"test/my-test-name.integ.snapshot\\" ] || (echo \\"No snapshot available for integration test 'my-test-name'. Run 'projen integ:my-test-name:deploy' to capture.\\" && exit 1)",
    },
    Object {
      "exec": "cdk8s synth --app \\"ts-node -P tsconfig.dev.json test/my-test-name.integ.ts\\" -o test/.tmp/my-test-name.integ/assert.cdk.out > /dev/null",
    },
    Object {
      "exec": "diff test/my-test-name.integ.snapshot/ test/.tmp/my-test-name.integ/assert.cdk.out/",
    },
  ],
}
`;

exports[`IntegrationTest adding an integration test 4`] = `
Object {
  "description": "update snapshot for all integration tests",
  "name": "integ:snapshot-all",
  "steps": Array [
    Object {
      "spawn": "integ:my-test-name:snapshot",
    },
  ],
}
`;

exports[`IntegrationTest adding an integration test 5`] = `
Object {
  "description": "Run tests",
  "name": "test",
  "steps": Array [
    Object {
      "exec": "jest --passWithNoTests --all --updateSnapshot",
    },
    Object {
      "spawn": "eslint",
    },
    Object {
      "spawn": "integ:my-test-name:assert",
    },
  ],
}
`;

exports[`IntegrationTest explicit name 1`] = `
Object {
  "description": "update snapshot for all integration tests",
  "name": "integ:snapshot-all",
  "steps": Array [
    Object {
      "spawn": "integ:foobar:snapshot",
    },
  ],
}
`;

exports[`IntegrationTest explicit name 2`] = `
Object {
  "description": "Run tests",
  "name": "test",
  "steps": Array [
    Object {
      "exec": "jest --passWithNoTests --all --updateSnapshot",
    },
    Object {
      "spawn": "eslint",
    },
    Object {
      "spawn": "integ:foobar:assert",
    },
  ],
}
`;
